### PR TITLE
Add Dockerfile and Docker quickstart guide

### PR DESCRIPTION
**Pull request by neufeldtech**
I've created a simple Dockerfile and guide to be considered for inclusion in this repo.
The goal in mind is to lower the entry barrier for users that want to quickly get started and explore this repo without having to install all of the projects dependencies (Caffe).

### CODE REVIEW

**Comment by yahoocla** [comment]
Thank you for submitting this pull request, however I do not see a valid CLA on file for you. Before we can merge this request please visit https://yahoocla.herokuapp.com/ and agree to the terms. Thanks! :smile:

**Comment by yahoocla** [comment]
CLA is valid!

**Comment by jay-mahadeokar** [comment]
Thanks for adding docker support. Can you also add steps / links on docker installation (if possible for Mac)? I will try it out. 
In case there are lot of steps (for different OS) I think we could add a [wiki page](https://github.com/yahoo/open_nsfw/wiki) for docker setup and point to it from this Readme, so that the Readme remains compact. What do you think?

**Comment by neufeldtech** [comment]
I've added Docker Installation links for Windows, Mac and Ubuntu. I have tested the getting started guide on Mac OSX and Ubuntu but do not have a Windows machine for testing at this time.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,4 @@
+FROM kaixhin/caffe
```
Comment (jay-mahadeokar) [inline Dockerfile]: @neufeldtech Is this the official caffe docker package? There is also another guide to [docker](https://github.com/yahoo/open_nsfw/issues/7), lets consolidate it in this PR.

**Comment by loretoparisi** [comment]
@neufeldtech it seems we did both the same! Please take a look here also if it can help https://github.com/yahoo/open_nsfw/issues/7 so that we will have just one PR 👍

**Comment by neufeldtech** [comment]
@jay-mahadeokar I've updated the README to build the docker image directly from the official BVLC caffe repo now.

@loretoparisi I like your approach! I've collapsed a few steps of your instructions into fewer commands. Take a look at what I've got so far here: [68040e6](https://github.com/yahoo/open_nsfw/pull/2/commits/68040e6b87c92f0f2e3c4d69dc599efe852b3e51)
Note: I opted not to include how to build the GPU image because I could not get it to work. What specific requirements/dependencies would someone need in order for this to work with the GPU image?

**Comment by jay-mahadeokar** [comment]
Thanks for adding docker support @neufeldtech  @loretoparisi.

**Comment by loretoparisi** [comment]
@neufeldtech basically you need to install on the host machine the nvidia `CUDA` 8.0 drivers, what we could do is

1) use `caffe` build through the `caffe:gpu` directly on the host machine:

```
$ docker build -t caffe:gpu ./
```

we verify

```
$ docker run -ti caffe:gpu caffe --version
libdc1394 error: Failed to initialize libdc1394
caffe version 1.0.0-rc3
```

This will download nvdia CUDA drivers and install it

```
$ nvidia-smi
Wed Oct  5 10:36:29 2016       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.44                 Driver Version: 367.44                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GT 620      Off  | 0000:01:00.0     N/A |                  N/A |
| 52%   30C   P12    N/A /  N/A |    247MiB /   962MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0                  Not Supported                                         |
+-----------------------------------------------------------------------------+
```

Note that we have one GPU (a very cheap `GeForce GT 620` here) addressed at `0`


And our container is up & running

```
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS                     PORTS                                            NAMES
443a7addcdff        caffe:gpu                     "caffe --version"        2 minutes ago       Exited (0) 2 minutes ago                                                    naughty_hamilton
```

2) Using the [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) container. In this way it would be very easy to add optional support to gpu that will work linux machines (Ubuntu/CentOS) with a supported CUDA 8.0 gpu (the problem here is that CUDA 8.0 moved to the new `Pascal` architecture, so it needs new graphic cards drivers on old gpus, etc.

In this case we could do

```
# Install nvidia-docker and nvidia-docker-plugin
wget -P /tmp https://github.com/NVIDIA/nvidia-docker/releases/download/v1.0.0-rc.3/nvidia-docker_1.0.0.rc.3-1_amd64.deb
sudo dpkg -i /tmp/nvidia-docker*.deb && rm /tmp/nvidia-docker*.d
```

and verify again that the nvidia docker container is getting the gpu as well

```
$ nvidia-docker run --rm nvidia/cuda nvidia-smi
Wed Oct  5 08:39:16 2016       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.44                 Driver Version: 367.44                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GT 620      Off  | 0000:01:00.0     N/A |                  N/A |
| 52%   31C   P12    N/A /  N/A |    247MiB /   962MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID  Type  Process name                               Usage      |
|=============================================================================|
|    0                  Not Supported                                         |
+-----------------------------------------------------------------------------+
```

ok it seems so, 

```
./nvidia-docker-run --volume=$(pwd):/workspace --rm -it caffe:gpu bash
```

so that we will have `caffe` running with the gpu on the host machine:

```
root@1bd4c3073ef0:/workspace# python ./classify_nsfw.py --model_def nsfw_model/deploy.prototxt --pretrained_model nsfw_model/resnet_50_1by2_nsfw.caffemodel anderson_210.jpg
```

If we can this typical error

```
F1005 08:48:48.429678    16 cudnn_conv_layer.cpp:52] Check failed: error == cudaSuccess (35 vs. 0)  CUDA driver version is insufficient for CUDA runtime version
```

it means that we have to update the `CUDA` driver as stated above.

**Comment by jay-mahadeokar** [comment]
I think it is best to have a wiki page with docker information and documentation.   We can move all the documentation to that page, so that the Readme itself remains concise.  @loretoparisi we can also move the existing docker documentation in Readme to that wiki page.

**Pull request merged**